juju/model.py

```diff
@@ -0,0 +1,172 @@
+
+
+class LazyMapping(Mapping, ABC):
+    @abstractmethod
+    def _load(self):
+        raise NotImplementedError()
+
+    @property
+    def _data(self):
+        if not hasattr(self, '_lazy_data'):
+            self._lazy_data = self._load()
+        return self._lazy_data
+
+    def __contains__(self, key):
+        return key in self._data
+
+    def __len__(self):
+        return len(self._data)
+
+    def __iter__(self):
+        return iter(self._data)
+
+    def __getitem__(self, key):
+        return self._data[key]
+
+
+class RelationMap(LazyMapping):
+    """Map of relation names to lists of Relation instances."""
+    def __init__(self, relation_names, backend, entity_cache):
+        self._relation_names = relation_names
+        self._backend = backend
+        self._entity_cache = entity_cache
+
+    def _load(self):
+        data = {}
+        for relation_name in self._relation_names:
+            relations = data[relation_name] = []
+            for relation_id in self._backend.relation_ids(relation_name):
+                relations.append(Relation(relation_name, relation_id, self._backend, self._entity_cache))
+        return data
+
+
+class Relation:
+    def __init__(self, relation_name, relation_id, backend, entity_cache):
+        self.relation_name = relation_name
+        self.relation_id = relation_id
+        self._backend = backend
+        self.data = RelationData(self.relation_name, relation_id, self._backend, entity_cache)
+
+    @property
+    def apps(self):
+        return [unit.app for unit in self.units]
+
+    @property
+    def units(self):
+        return list(self.data.keys())
+
+
+class RelationData(LazyMapping):
+    def __init__(self, relation_name, relation_id, backend, entity_cache):
+        self.relation_name = relation_name
+        self.relation_id = relation_id
+        self._backend = backend
+        self._entity_cache = entity_cache
+
+    def _load(self):
+        data = {}
+        for unit_name in self._backend.relation_list(self.relation_id):
+            unit = self._entity_cache.get(Unit, unit_name)
+            data[unit] = RelationDataBag(self.relation_id, unit, self._backend)
+        # Juju's relation-list doesn't include the local unit(s), even though they are part of
+        # the relation. Technically, you can also call relation-get for your peers' data, but
+        # we don't want to support that, so we only manually add this local unit.
+        local_unit = self._entity_cache.get(Unit, self._entity_cache.local_unit_name)
+        data[local_unit] = RelationDataBag(self.relation_id, local_unit, self._backend)
+        return data
+
+
+class RelationDataBag(LazyMapping):
+    def __init__(self, relation_id, unit, backend):
+        self.relation_id = relation_id
+        self.unit = unit
+        self._backend = backend
+
+    def _load(self):
+        return self._backend.relation_get(self.relation_id, self.unit.name)
```

Comment:
We're probably going to have trouble with this logic when we try to change settings. Let's leave this as-is for now so we can get this PR which is mostly ready out of the way. In the future we'll most likely need to tune this further so we can tell what is changing and when. Anyway, not for now.